### PR TITLE
Fix loss scaling in Trainer during final step of gradient accumulation

### PR DESCRIPTION
**Pull request by andrewtran117**
Fix for bug in issue (#38837)

# What does this PR do?
This PR fixes a bug in the Trainer.training_step method where the loss is always divided by gradient_accumulation_steps, even on the final accumulation cycle of an epoch, which may contain fewer batches. This leads to an artificially small loss during the last optimizer steps. For example, the following buggy output is provided by @hutaiHang:
"
{'loss': 0.9984, 'grad_norm': 36.21815490722656, 'learning_rate': 0.0, 'epoch': 0.4}
{'loss': 0.9984, 'grad_norm': 36.21815490722656, 'learning_rate': 0.0, 'epoch': 0.8}
{'loss': 0.4992, 'grad_norm': 18.10907745361328, 'learning_rate': 0.0, 'epoch': 1.0} <-- The problem!
{'loss': 0.9984, 'grad_norm': 36.21815490722656, 'learning_rate': 0.0, 'epoch': 1.4}
"

In this PR, I provide a fix by tracking the number of batches accumulated before each optimizer step. Rather than dividing the loss by 'gradient_accumulation_steps,' the loss is now divided by the true number of accumulated steps. 

Fixes #38837
Loss is incorrectly scaled in Trainer during the last step with gradient accumulation when the final batch is smaller than accumulation steps.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ x ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Tagging same people from original issue: @zach-huggingface @SunMarc

### CODE REVIEW

**Comment by andrewtran117** [comment]
Thanks for the feedback! @hutaiHang
